# Ethan Pence
# CPSMA-3933-01, 12:30 PM - 1:45 PM
# Lab 5

First, I will find the analytical equation.

f(0) = 1, f(1) = 3

r^n = 7 * r^(n - 1) - 10 * r^(n - 2)

r^2 = 7r - 10

r^2 - 7r + 10 = 0

(r - 5)(r - 2)

r1 = 2, r2 = 5

f(0) = 1 = c_1 * (2)^0 + c_2 * (5)^0

c_1 + c_2 = 1

f(1) = 3 = c_1 * (2)^1 + c_2 * (5)^1

2c_1 + 5c_2 = 3

-2(c_1 + c_2 = 1) = ( -2c_1 - 2c_2 = -2 )

(-2c_1 - 2c_2) + (2c_1 + 5c_2) = -2 + 3

3c_2 = 1

c_2 = 1/3

c_1 = 2/3

f(n) = 2/3 * 2^n + 1/3 * 5^n


Below is a function made from the equation. (made to use ints)

In [45]:
def analytic_equation(n):
    return (5**n + 2**(n+1)) // 3

print(analytic_equation(6))

5251


Below is another way of solving the equation using bottom up dynamic programming.

In [46]:
a_n_1 = 3
a_n_2 = 1

a_n = 0

for n in range(5):
    a_n = 7 * a_n_1 - 10 * a_n_2
    
    a_n_2 = a_n_1
    a_n_1 = a_n
    
print(a_n)

5251


Below is an example solving the equation using top down dynamic programming and memoization.

In [47]:
def recursive(n, m_array):
    if m_array[n] is not None:
        return m_array[n]
    
    if n == 0:
        return 1

    if n == 1:
        return 3
    
    solution = 7 * recursive(n - 1, m_array) - 10 * recursive(n - 2, m_array)
    
    m_array[n] = solution
    
    return solution

def call(n):
    m_array = [None] * (n + 1)
    return recursive(n, m_array)
    
print(call(6))

5251


All three functions achieved the same result of 5251 for f(6).

The analytic function is the fastest in terms of computation time, however, it is also the most difficult to create.

The bottom up method is both simple to create and the second fastest.

The top down method is roughly equal in difficulty of creation with the bottom up method. However, the top down method can easily run into stack overflow errors at larger values of n.

All of the methods are able to calculate n = 1000 almost instantly.